### ISSUE TITLE
[api-extractor] support ambient types

### ISSUE THREAD

**Issue by HolgerJeromin** [Microsoft/web-build-tools#1176]
I have a quite huge library which `.d.ts` API file I want to "filter" with api extractor. Only `@public` stuff should survive in the `.d.ts` file I give to customers. 

The code consists of code which uses `modules="none"`. 
When I call API extractor on its `.d.ts` file I get this error:
```
Error: The entry point file does not appear to have any exports:
xy/hmi.d.ts
Note that API Extractor does not yet support libraries consisting entirely of ambient types.
```
Did i miss something? Is there a workaround for this?
Thanks a lot!

Minimal example:
```ts
namespace Hmi {
    console.log('Hello World');

    /** @public  */
    export module Log {
        export function error(message: string | object): void {         
            if (console && console.error) {            
                    console.error(message);             
            }
        }
    }
    /**  @internal */
    export module System {
       export class PrivateStuff {
            public constructor(){
                console.log('private stuff');
            }
        }
    }    
}
```

tsconfig.json
```json
{
    "$schema": "http://json.schemastore.org/tsconfig",
    "compilerOptions": {
       "module": "none",
       "target": "es5",
       "declaration": true
    },
    "files": [
       "source.ts"
    ],
    "include": [
       
    ]
 }
```

**Comment by octogonz** [comment]
I haven't seen this configuration before. We should support it, but it hasn't ever been tested as far as I know. I won't have time to address it until after AE7 ships. If you want to contribute a PR that would be cool also.

But I wonder... could you work around this issue by changing the tsconfig.json file that is used when running API Extractor? It can be different from what you use for compiling. (The api-extractor.json file has a `overrideTsconfig` setting for this.)

**Comment by HolgerJeromin** [comment]
> We should support it

In the sense of: "this should be possible" not "this seems like a bug", right?

> could you work around this issue by changing the tsconfig.json file that is used when running API Extractor? 

I did not found this in https://api-extractor.com/pages/setup/config_file/ but I will try next week. Thanks for the hint!

**Comment by octogonz** [comment]
> In the sense of: "this should be possible" not "this seems like a bug", right?

Ideally API Extractor should support every exotic usage of TypeScript imaginable, even if only a few people are doing that. But given the limited developer resources that we have, we realistically need to draw a line somewhere for bugs, enhancements, stuff we don't have time to support, and stuff we cannot support because it introduces too much complexity. These lines are subjective, and should be driven by community demand.

As an example, supporting `export * from` was a huge amount of work, but I invested in it because lots of people kept asking for it, and there weren't easy workarounds in some cases.

**Comment by justintime4tea** [comment]
I'm confused on how to do `export * from` if you're talking about reexporting in a file because you use relative paths everywhere else. I can't for the life of me generate a roll-up and all I get is `export { }`.

**Comment by octogonz** [comment]
Can you share a repro branch? In the previous discussion it seemed that ` "module": "none"` was causing the trouble.

**Comment by justintime4tea** [comment]
I may have figured it out but I don't know if it's a workaround of mine or if it's the right way to go. When using `export default class Main` (for example) and then reexporting in index.ts using `export * from './Main'` it would not show up in the ts-template.d.ts file in my dist. When removing default from the export in Main.ts it then rolled it up into ts-template.d.ts (along with the others which I had to remove default from as well).

Ultimately I'd like to move away from the path based importing all-together but it seamed like I still had a lot to learn before making the jump. I'd prefer to just do non-relative imports honestly.

Apologies if this is a waste of your time. I am extremely new to Typescript but am extremely experienced with JavaScript, Java, C#, Python, etc... I'm really excited about Typescript so I want to do things right.

**Repo (rollup works with out default exports):**
https://github.com/justintime4tea/ts-template

**Failing rollup (with default exports rollup not working):**
https://github.com/justintime4tea/ts-template/tree/default-exports-no-rollup

**Comment by octogonz** [comment]
`export default` is not a good practice, because the compiler isn't able to validate the name of the thing you are importing. For example you can accidentally have `import Dog from './cat'` and then get a bunch of confusing errors about your dog meowing instead of barking. Whereas if you remove `default` then the import becomes `import { Dog } from './cat'` and you'll get an error right away when it can't find an export named `Dog`.

**Comment by octogonz** [comment]
> **Failing rollup (with default exports rollup not working):**
> https://github.com/justintime4tea/ts-template/tree/default-exports-no-rollup

This is failing because `export * from './Main';` ignores the default export in **Main.ts**.  That construct only re-exports named exports.  The class is called `Main` locally inside that file, but in the exports table, it is unnamed.

Thus API Extractor is doing the right thing by omitting it from the rollup.

Your https://github.com/justintime4tea/ts-template branch is the right way to go.

**Comment by justintime4tea** [comment]
Well then that sounds great! I will do that instead and things are now working. Thanks a lot @pgonzal ! 

Now for my journey on to NOT using relative paths ha 🥇

**Comment by octogonz** [comment]
Your relative paths are not a problem.

It's bad to create a library where people need to use "path-based imports" like this:
```ts
import { Something } from 'some-library/folder1/folder2/Something';
```
That's bad because the internal folder structure (an implementation detail) has become part of the public API contract for the NPM package `some-library`.

But it's totally fine to do this:
```ts
import { Something } from './folder1/folder2/Something';
```
This is also okay:
```ts
export { Something } from './folder1/folder2/Something';
```

Here we are referring to other source files within the same NPM package as ourselves, so these paths are local.  When API Extractor makes the rollup, these "relative paths" will get discarded.  Our NPM package will still have a single entry point (**ts-template/src/index.ts** in your example).

**Comment by justintime4tea** [comment]
I agree that internal to the library itself path based isn't terrible but making a user of a library use paths is a recipe for disaster. On board 100%. Eloquently put. 

Thanks again for the help! Very insightful and I appreciate your work on these projects!

**Comment by andykais** [comment]
@octogonz @justintime4tea what about tree shaking? Especially with frontend libraries, sometimes it is imperative to have path based imports.
E.g. [react material-ui](https://github.com/mui-org/material-ui) or even [lodash](https://lodash.com/)

**Comment by octogonz** [comment]
@andykais https://github.com/microsoft/web-build-tools/issues/664 tracks support for path-based imports.  We do want to support it.  It's a nontrivial feature, though.  How do you represent them in the docs?  For the .d.ts rollup, what do you do with an unexported type that is used by two different entry points?  These problems are certainly solvable.  But the API Extractor maintainers have never needed to use path-based imports in any of our APIs, and we have limited resources.  It would be a great feature for someone from the community to help out with.

> what about tree shaking?

I thought the whole point of tree-shaking was to eliminate dead code, i.e. to automate the optimization that is otherwise manually performed by managing path-based imports.

**Comment by andykais** [comment]
> automate the optimization that is otherwise manually performed by managing path-based imports.
ah, it appears things have changed. Since webpack 2 apparently, path based imports are not necessary for tree shaking. Thanks for the tip!

I will probably just assume developers are using a decently new bundler to get tree shaking without path based imports. I am unblocked!

**Comment by MarkDuckworth** [comment]
+1 vote for supporting ambient types.
